matrix_client/crypto/olm_device.py

```diff
@@ -0,0 +1,56 @@
+import logging
+
+import olm
+from canonicaljson import encode_canonical_json
+
+from matrix_client.checks import check_user_id
+
+logger = logging.getLogger(__name__)
+
+
+class OlmDevice(object):
+    """Manages the Olm cryptographic functions.
+
+    Has a unique Olm account which holds identity keys.
+
+    Args:
+        api (MatrixHttpApi): The api object used to make requests.
+        user_id (str): Matrix user ID. Must match the one used when logging in.
+        device_id (str): Must match the one used when logging in.
+    """
+
+    def __init__(self, api, user_id, device_id):
+        self.api = api
+        check_user_id(user_id)
+        self.user_id = user_id
+        self.device_id = device_id
+        self.olm_account = olm.Account()
+        self.olm_account.create()
+        logger.info('Initialised Olm Device.')
+
+    def sign_json(self, json):
+        """Signs a JSON object.
+
+        As specified, this is done by encoding the JSON object without ``signatures`` or
+        keys grouped as ``unsigned``, using canonical encoding.
+
+        Args:
+            json (dict): The JSON object to sign.
+
+        Returns:
+            The same JSON object, with a ``signatures`` key added. It is formatted as
+            ``"signatures": ed25519:<device_id>: <base64_signature>``.
+        """
+        signatures = json.pop('signatures', {})
+        unsigned = json.pop('unsigned', None)
+
+        signature_base64 = self.olm_account.sign(encode_canonical_json(json))
+
+        key_id = 'ed25519:{}'.format(self.device_id)
+        signatures.setdefault(self.user_id, {})[key_id] = signature_base64.decode()
```

Comment:
FYI the code is pretty much taken from the doc https://matrix.org/docs/spec/appendices.html#id7

Are you still sure about this? I was confused by `setdefault` too when I discovered it by reading the above code, and now I find myself abusing it :)

Reply:
Even knowing how `setdefault` works, I still find the two-line version immensely more readable. I'd prefer to switch the doc to two lines also tbh.

Reply:
Well, I don't, because knowing how `setdefault` works I find it quite confusing to use it like that, ignoring its return value. Here's an idea to have the best of both worlds:
```signatures_user_id = signatures.setdefault(self.user_id, {})
signatures_user_id[key_id] = signature_base64.decode()```
Not sure the name `signatures_user_id` is good at clarifying anything though...

Reply:
Let's not introduce a temporary variable for sure. The purpose of `setdefault` as I understand it is to set a dictionary value without first having to check for existence. I do believe my suggestion is more readable, but it's not something worth it to me to actually present to multiple people and prove it, so let's leave it as is. :)